Fix version comparison with minor and patch level requirements

ChristophWurst opened the pull request:
If an app requires a specific minor or path level server version,
the version_compare prevented the installation as only the major
version had been compared and that checks obviously returns `false`.


```
php > var_dump(version_compare('13', '13.0.1', '>='));
bool(false)
php > var_dump(version_compare('13.0.2', '13.0.1', '>='));
bool(true)
```

Now the full version is used for comparison, making it possible to
release apps for a specific minor or patch level version of Nextcloud.

This is a quick fix for discussion. There are probably some tests that will fail now.

cc @rullzer @nickvergessen as discussed.

ChristophWurst committed:
Fix version comparison with minor and patch level requirements

If an app requires a specific minor or path level server version,
the version_compare prevented the installation as only the major
version had been compared and that checks obviously returns `false`.

Now the full version is used for comparison, making it possible to
release apps for a specific minor or patch level version of Nextcloud.

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

nickvergessen commented on lib/private/App/AppStore/Fetcher/AppFetcher.php:
```diff
@@ -84,10 +83,10 @@ protected function fetch($ETag, $content) {
+						// Version is bigger or equals to the minimum version of the app
+						version_compare($ncVersion, $version->getMinimumVersion(), '>=')
+						// Version is smaller or equals to the maximum version of the app
+						&& version_compare($ncVersion, $version->getMaximumVersion(), '<=')
```
Ah here is what breaks, all apps currently have ` max-version="13"`
So we need to keep the major only check for the max version.

ChristophWurst replied:
That's disputable. I think if someone specifies the full version requirement like `13.0.1` (for whatever reason) we should do the full check. But of course, if max is set to `13`, we should allow any version of Nextcloud 13.

MorrisJobke replied:
Couldn't we align the comparison to the same amount of digits? Then use the `<=` and `>=` operators and it would be fine. If max version is `13.0.2` we would compare against the full version if it is `13` we would only use the major version.

ChristophWurst commented:
I've added a new version comparison class to compare app requirements. I chose to move the logic into a new class because it's likely to be used in other places too and because it makes testing a lot easier. Ref https://github.com/nextcloud/server/pull/9024#discussion_r177977853.

There's one issue: the existing test cases fail now, because they use a `11.0.0.2` server version for testing. Is that even a valid server version? Should I adjust the test or my logic? In case of the latter: what would be the semantics of x.x.x.x version comparisons? cc @MorrisJobke @nickvergessen

MorrisJobke commented:
> There's one issue: the existing test cases fail now, because they use a 11.0.0.2 server version for testing. Is that even a valid server version? Should I adjust the test or my logic? In case of the latter: what would be the semantics of x.x.x.x version comparisons? cc @MorrisJobke @nickvergessen

Yes it is. The 4th digit is our internal version number to be able to trigger DB migrations ;) But they should now be used outside of our server core. So the info.xml should at most hold 3 parts.

ChristophWurst commented:
> Yes it is. The 4th digit is our internal version number to be able to trigger DB migrations ;) But they should now be used outside of our server core. So the info.xml should at most hold 3 parts.

Fixed. Please see the test data to verify I implemented the compatibility logic correctly :wink:

ChristophWurst commented on lib/private/App/AppStore/Fetcher/AppFetcher.php:
```diff
@@ -83,12 +89,13 @@ protected function fetch($ETag, $content) {
+					$min = $version->getMinimumVersion();
+					$max = $version->getMaximumVersion();
+					$minFulfilled = $this->compareVersion->isCompatible($ncVersion, $min, '>=');
+					$maxFulfilled = $max !== '' &&
```
FYI this is merely for backwards compatibility. Nowadays apps *have to* specify their min *and* max requirements, thus this check is not necessary anymore. However, the test data contains apps that have no max-requirement set and thus this code would fail. Since I'm not 100% sure if the app store would distribute these apps, I took the safe path and added this check defensively.

MorrisJobke commented on lib/private/App/CompareVersion.php:
```diff
@@ -0,0 +1,97 @@
+	public function isCompatible(string $actual, string $required,
+		string $comparator = '>='): bool {
+
+		if (!preg_match(self::REGEX_SERVER, $actual)) {
```
Wouldn't it make sense to do then only a check agains `REGEX_MAJOR_MINOR_PATCH` because it is excluded below anyways (in the else branch)?

ChristophWurst replied:
Below only `$required` is checked, not `$actual`. And the server version is in x.x.x(.x) format, hence the special regex with the optional forth-level version.

MorrisJobke commented on lib/private/App/AppStore/Fetcher/AppFetcher.php:
```diff
@@ -83,12 +89,13 @@ protected function fetch($ETag, $content) {
+					$ncVersion = $this->getVersion();
+					$min = $version->getMinimumVersion();
+					$max = $version->getMaximumVersion();
+					$minFulfilled = $this->compareVersion->isCompatible($ncVersion, $min, '>=');
```
Also catch the exceptions and log them - otherwise this gets messy and user complain about broken app pages and stuff like that.

ChristophWurst replied:
> Also catch the exceptions and log them - otherwise this gets messy and user complain about broken app pages and stuff like that.

What shall we do in case of a thrown exception? Ignore the app? Install anyway? IMO this one is unexpected for normal use and is only thrown if the class is used wrongly.

MorrisJobke replied:
It should ignore the app - would be horrible if you have one broken app installed and the app management would not work anymore.

rullzer replied:
* catch it => Log it
* And remove from the app list so people don't even see it (so can't install)

ChristophWurst replied:
FYI: while we could add that (IMO ugly) check, it's not actually possible that an app would have an invalid version requirement at the moment because it's validated with the info.xml schema:
```
    <xs:simpleType name="version">
        <xs:restriction base="limited-string">
            <xs:pattern value="[0-9]+(\.[0-9]+){0,2}"/>
        </xs:restriction>
    </xs:simpleType>
```

MorrisJobke committed:
Log possible exception caused by wrong version number

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke reviewed (approved):
Fine now 👍

MorrisJobke commented:
@ChristophWurst Please have a final look at my changes 😉

ChristophWurst reviewed (commented):
:elephant:

MorrisJobke commented:
Dumping the autoloader is missing

rullzer committed:
Bump autoloader

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>
